design/cli-install-changes.md

```diff
@@ -0,0 +1,426 @@
+
+The user will start up the Velero server on a cluster by using the command `velero config server`. This will create the Velero deployment resource with default values or values overwritten with flags, create the Velero CRDs, and anything else that is not specific to plugins or BSL/VSL.
+
+The Velero server will start up, verify that the deployment is running, that all CRDs were found, and log a message that it is waiting for a BSL to be configured. at this point, other operations, such as configuring restic, will be allowed. Velero should keep track of its status, ie, if it is ready to create backups or not. This could be a field `ServerStatus` added to `ServerStatusRequest`. Possible values could be [ready|waiting]. "ready" would mean there is at least 1 valid BSL, and "waiting" would be anything but that.
+
+When adding/configuring a BSL or VSL, we can approach it in one of two ways:
+1) at the time of creating a location, verify if there is a valid, corresponding plugin. If there isn't, don't allow creation.
+2) allow creating locations, and continuously verify if there is a corresponding, valid plugin. When a valid match is found, mark the BSL/VSL as "ready". This would require adding a field to the BSL/VSL to keep track of its status, possibly: [ready|waiting].
+
+With the first approach: the server would transition into "ready" (to create backups) as soon as there is one BSL. It would require a set sequence of actions, ie, first install the plugin, only then the user can successfully configure a BSL.
+
+With the second approach, the Velero server would continue looping and checking all existing BSLs for at least 1 with a "ready" status. Once it found that, it would set itself to "ready" also.
+
+Another new behavior that must be added: the server needs to identify when there no longer exists a valid BSL. At this point, it should change its status from "ready" to one that indicates it is not ready, maybe "waiting". With the first approach above, this would mean checking if there is still at least one BSL. With the second approach, it would require checking the status of all BSLs to find at least one with the status of "ready".
+
+As it is today, a valid VSL would not be required to create backups, unless the backup included a PV.
+
+To make it easier for the user to identify if their Velero server is ready to create backups or not, a `velero status` command should be added. This issue has been created some time ago for this purpose: https://github.com/vmware-tanzu/velero/issues/1094.
+
+## Alternatives Considered
+
+It seems that the vast majority of tools document their usage with `kubectl` and `yaml` files to install and configure their Kubernetes resources. Many of them also make use of Helm, and to a lesser extent some of them have their own CLI tools. 
```

Comment:
Should we consider maintaining static YAML files?

I know we moved away from this in the past because they were a maintenance burden, but with the changes to plugins and BSL/VSL, the content in them is now more factored out, meaning we wouldn't have to maintain so many provider-specific permutations.

Reply:
I would prefer to have these static files, but generated at release time. To have a per release set of these static yaml files.

Reply:
After perusing other projects, I'd say so, since it seems very common to use yaml for operations. Also, I thought it'd be useful to keep example yaml files and kustomize files, so we can maintain the yaml files and add a kustomize files to serve as example. Killing two birds with one stone.

Reply:
Yeah, the discussion about YAML vs...well, anything else is on-going in the larger community, but YAML is really what everything ends up boiling down to.

The biggest question that remains for me is how the plugin overlays/patches get generated in a sustainable way. I 100% agree with @ashish-amarnath that we should have these generated via tooling on release, I'm just not sure what tooling would look like for the overlays yet. My current best idea is that we'd have a running KinD cluster with the relevant plugin installed and generate a diff of the YAML, but that seems very heavyweight.

Generating the base file(s) wouldn't really be too far off from what we have in our install dry run now, at least. The biggest difference, I think, is generating them such that they get hosted somewhere convenient like with https://projectcontour.io/quickstart/contour.yaml. Maybe a GitHub action & bot tosses them into directory in `_site/docs/<version>` when the release is cut?

Reply:
procedural question - do we want to continue the YAML discussion in a separate issue and not block this CLI work on it? Since we're not immediately dropping `velero install`, we continue to have a viable path for users, and I do think we need to do some more discussing on kustomize/static YAML/plugin patches/etc. before we know exactly what our path forward is there.

Reply:
Yes, that'd be great imo.

Reply:
> do we want to continue the YAML discussion in a separate issue and not block this CLI work on it? 

Yeah, let's do that.

Reply:
https://github.com/vmware-tanzu/velero/issues/2405